### PR TITLE
Allow SecurityContext and PodSecurityContext to be reset to default

### LINKED ISSUE #741: Override does not patch with null values
## Describe the bug

Create a RMQ with manifest like so:

```yaml
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: sample
  namespace: rabbitmq-system
spec:
  image: rabbitmq:3.8.17-management
  override:
    statefulSet:
      spec:
        template:
          spec:
            securityContext: null
            containers: []
            initContainers:
            - name: setup-container
               securityContext: null
```

**Expected behavior**

Deployed statefulSet pod spec should have securityContext set to `null` or not set (empty). However, both `securityContext: null` is  omitted after the manifest is saved (if you get the RMQ manifest after creation, you don't see `securityContext: null` in the manifest).

Setting fields to `null` is a documented way of deleting a field for [strategic merge patch ](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#maps). When applying the same patch but using `kubectl patch` directly, securityContext can be set to null succesfully. 

Example patch:
```yaml
spec:
 template:
   spec:
     securityContext: null
     initContainers:
     - name: setup-container
       securityContext: null
```
applied by running `k patch sts sample-server --patch "$(cat patch.yaml)"`

## Version and environment information
- RabbitMQ: 3.8.17
- RabbitMQ Cluster Operator: latest main
- Kubernetes: [e.g. 1.18.0]
- Cloud provider or hardware configuration:

### PR DESCRIPTION
**Pull request by coro**
This closes #741 and closes #746 

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

## Summary Of Changes
If an override is provided with the SecurityContext of either the rabbitmq container or init container set to `{}`, this ensures the resultant STS uses the default securityContext of the container runtime.

## Additional Context
Unlike #746, this preserves the validation of override types, rather than failing at reconcile time if a user provides an unknown field.

This only works for securityContext right now; the only other fields I think this affects are:
`topologySpreadConstraints`
`resources`

I don't think they're a high priority to make this work for, but I could add them in if needed.

## Local Testing

Please ensure you run the unit, integration and system tests before approving the PR.

To run the unit and integration tests:

```
$ make unit-tests integration-tests
```

You will need to target a k8s cluster and have the operator deployed for running the system tests.

For example, for a Kubernetes context named `dev-bunny`:
```
$ kubectx dev-bunny
$ make destroy deploy-dev
# wait for operator to be deployed
$ make system-tests
```

### CODE REVIEW

File: controllers/rabbitmqcluster_controller_test.go [hunk 1]
```diff
@@ -957,6 +958,28 @@ var _ = Describe("RabbitmqClusterController", func() {
 				return c.Image
 			}, 3).Should(Equal("my-great-image-2"))
 		})
+
+		It("can reset a controller-set field to default", func() {
```
Comment (ChunyiLyu) [inline controllers/rabbitmqcluster_controller_test.go:962]: Should we make the test description more specific here? Our controller is only treating pod and container SecurityContext as special cases, and it currently does not handle any other fields

File: internal/resource/statefulset.go [hunk 2]
```diff
@@ -258,6 +259,16 @@ func patchPodSpec(podSpec, podSpecOverride *corev1.PodSpec) (corev1.PodSpec, err
 		sortVolumeMounts(patchedPodSpec.Containers[0].VolumeMounts)
 	}
 
+	if podSpecOverride.SecurityContext != nil && reflect.DeepEqual(*podSpecOverride.SecurityContext, corev1.PodSecurityContext{}) {
```
Comment (ChunyiLyu) [inline internal/resource/statefulset.go:262]: I would suggest adding comments here with context about why we are handing securityContext as a special case. I can see myself getting confused in the future.

**Review by ChunyiLyu — APPROVED** [review]
LGTM. Just some minor suggestions on comments and tests descriptions.

**Review by ablease — APPROVED** [review]
LGTM. Nothing to add to Chunyi's suggestions. 👍

**Review by ChunyiLyu — APPROVED** [review]

**Pull request merged**